cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,137 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of queue.
+ *
+ * Implemented methods are not boxed, when using with primitive types.
+ */
+abstract class CQueue[@specialized(Int, Long) T] {
+  def first(): T
+  def +=(elem: T)
+  def deque(): T
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]].
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  def enqueueFirst(elem: T)
+}
+
+object CQueue {
+  /**
+   * Creates specialized [[CQueue]] that is a FIFO queue.
+   */
+  def fifo[@specialized(Int, Long) T]()(implicit impl: CQueueFactory[T]): FIFO[T]= {
+    impl.fifo()
+  }
+
+  /**
+   * Priority [[CQueue]] that uses default ordering (implicit) for `T`.
+   */
+  def naturalPriority[@specialized(Int, Long) T]()(
+    implicit order: Order[T],
+    impl: CQueueFactory[T]
+  ): CQueue[T] = {
+    impl.priority(Some(order))
+  }
+
+  /**
+   * Priority [[CQueue]] that uses provided `order`.
+   */
+  def priority[@specialized(Int, Long) T](order: Option[Order[T]])
+      (implicit impl: CQueueFactory[T]): CQueue[T] = {
+    impl.priority(order)
+  }
+
+  trait LowerPriorityImplicit {
+    private val factoryAny = new CQueueFactoryAny[AnyRef]()
+    implicit def factoryAny[T]: CQueueFactory[T] =
+      factoryAny.asInstanceOf[CQueueFactory[T]]
+  }
+
+  trait LowPriorityImplicit extends LowerPriorityImplicit {
+    implicit val factoryInt: CQueueFactory[Int] = new CQueueFactoryInt()
+  }
+
+  object Implicits extends LowPriorityImplicit
+
+}
+
+class CQueueAny[T, QueueType <: PriorityQueue[T]](protected val underlying: QueueType)
+    extends CQueue[T] {
+  def +=(elem: T): Unit = underlying.enqueue(elem)
+
+  def deque(): T = underlying.dequeue()
+
+  override def first(): T = underlying.first()
+
+  override def isEmpty: Boolean = underlying.isEmpty
+
+  override def clear(): Unit = underlying.clear()
+
+  override def size(): Int = underlying.size()
+}
+
+class CQueueInt[QueueType <: IntPriorityQueue](protected val underlying: QueueType)
+    extends CQueue[Int] {
+  override def +=(elem: Int): Unit = underlying.enqueue(elem)
+
+  override def deque(): Int = underlying.dequeueInt()
+
+  override def first(): Int = underlying.firstInt()
+
+  override def isEmpty: Boolean = underlying.isEmpty
+
+  override def clear(): Unit = underlying.clear()
+
+  override def size(): Int = underlying.size()
+}
+
+abstract class CQueueFactory[T] {
+  def fifo(): CQueue[T] with FIFO[T]
+  def priority(order: Option[Order[T]]): CQueue[T]
```

Comment:
how about not having Option, but having two overloaded methods:
1. def priority()
2. def priority(order: Order[T])
Just makes it clearer and there is no duplication of code because you have the two cases of Some() and None coded anyway